File: simulator/actions/dash.py
from functools import cache

from cachetools import cached
from cachetools.keys import hashkey

from ..actions.action_types import BonusAction, HasteAction
from ..actions.actoid import Actoid, ActoidFlags
import logging
from ..battle_map import Map, map_toggled_cache_with_key
from ..conditions import Conditions, is_affected_by_any
from ..threat_interfaces import AttackThreatModifier
from ..factory_interfaces import Factory

logger = logging.getLogger("Encounterra")


class DashFactory(Factory):
    def __init__(self, action_type, combatant):
        super().__init__()
        self.combatant = combatant
        self.action_type = action_type  # DASH, CUNNING_DASH, AGGRESSIVE

    def __str__(self):
        """
        Important for FSM building
        """
        return "DashFactory"

    def create_all(self, previous_action_in_dag=None):
        return [Dash(self)]

    def create(self, target):
        return Dash(self)


class Dash(AttackThreatModifier):
    def __init__(self, factory):
        AttackThreatModifier.__init__(self, ActoidFlags.IS_DASH | ActoidFlags.LOCATION_INDEPENDENT)
        self.name = "Dash"
        self.factory = factory
        if self.factory.action_type is BonusAction.AGGRESSIVE:
            self.actoid_flags |= ActoidFlags.IS_PRIORITY

    def __str__(self):
        if self.factory.action_type is BonusAction.CUNNING_DASH:
            return f"Cunning Dash {self.factory.combatant}"
        elif self.factory.action_type is HasteAction.HASTE_DASH:
            return f"Hasted Dash of {self.factory.combatant}"
        elif self.factory.action_type is BonusAction.AGGRESSIVE:
            return f"Aggressive movement of {self.factory.combatant}"
        return f"Dash of {self.factory.combatant}"

    def shorthand_str(self):
        if self.factory.action_type is BonusAction.CUNNING_DASH:
            return "Cunning Dash"
        elif self.factory.action_type is HasteAction.HASTE_DASH:
            return "Hasted Dash"
        elif self.factory.action_type is BonusAction.AGGRESSIVE:
            return "Aggressive movement"
        return "Dash"

    @map_toggled_cache_with_key(key=lambda self, **kwargs: hashkey(kwargs["movement_threat"], self.factory.combatant.movement, tuple(Map.get().get_combatant_position(self.factory.combatant).get()[0])))
    def calculate_threat(self, **kwargs):
        movement_threat = kwargs["movement_threat"]
        if self.factory.action_type is BonusAction.AGGRESSIVE:
            if len(movement_threat) - 1 > self.factory.combatant.movement:
                # We always want this to be used if the destination can't be reached, the moving towards an enemy part is always assumed
                return 1
            return -1
        baseline = -1 * movement_threat[min(int(self.factory.combatant.movement), len(movement_threat) - 1)]
        modified = -1 * movement_threat[min(int(self.factory.combatant.movement + self.factory.combatant.speed), len(movement_threat) - 1)]
        return max(0, baseline - modified)  # We're only interested in this if used defensively, we don't want it to play a role if used offensively

    def clear_cache(self):
        self.calculate_threat.cache_clear()
        #self.get_eligible_coords.cache_clear()

    def calculate_threat_for_attack(self, combatant, attack, *args, **kwargs):
        return 0  # TODO do the distance mod here

    #@map_toggled_cache_with_key(key=lambda self, distances, shortest_paths: hashkey(self.factory.name, tuple(Map.get().get_combatant_position(self.factory.combatant).get()[0])))
    def get_eligible_coords(self, distances, shortest_paths):
        battle_map = Map.get()
        if is_affected_by_any(self.factory.combatant, Conditions.GRAPPLED, Conditions.GRAPPLING, Conditions.RESTRAINED, Conditions.SWALLOWED):
            return None
        # if self.factory.combatant.movement > 0:
        return battle_map.get_all_accessible_coords(shortest_paths, self.factory.combatant)
        # return [tuple(battle_map.get_combatant_position(self.factory.combatant).get()[0])]

File: c++/actions/dash.cpp
#include "actions/dash.hpp"
#include "core/battle_map.hpp"
#include "core/combatant.hpp"
#include <algorithm>
#include <vector>

namespace enc
{
  std::vector<std::shared_ptr<Actoid>> DashFactory::createAll(void *previousActionInDag)
  {
    return {std::make_shared<Dash>(static_cast<ActoidFactory &>(*this))};
  }

  std::shared_ptr<Actoid> DashFactory::create(void *target) { return std::make_shared<Dash>(static_cast<ActoidFactory &>(*this)); }

  std::optional<CoordVector>
  Dash::getEligibleCoords(const blaze::DynamicVector<int> &distances, const blaze::DynamicMatrix<Coord> &shortestPaths)
  {
    // Mirrors Python actions.dash.Dash.get_eligible_coords. Dashing is location independent, so it can be taken
    // from ANY square the combatant can reach this turn (this is what lets a fleeing combatant Disengage, move
    // far, then Dash). It is unavailable while Grappled/Grappling/Restrained/Swallowed.
    Combatant *combatant = _factory.getCombatant();
    if(combatant->getSwallower() != nullptr
       || combatant->isAffectedByAny({Conditions::GRAPPLED, Conditions::GRAPPLING, Conditions::RESTRAINED}))
      {
        return std::nullopt;
      }
    return BattleMap::getInstance().getAllAccessibleCoords(shortestPaths, *combatant);
  }

  std::string Dash::toString() const
  {
    return _abilityType == AbilityType::CUNNING_DASH ? "Cunning Dash" : "Dash";
  }

  double Dash::calculateThreat(const Kwargs &kwargs)
  {
    // Mirrors Python actions.dash.Dash.calculate_threat. The planner threads the cumulative-threat-along-path
    // array it computed for this sequence's destination in via the "movementThreat" kwarg (only dash actoids
    // ask for it, so the common threat path never pays for building it). movementThreat[i] is the projected
    // threat (negative) of standing i cells along the path; index 0 is the current square.
    auto it = kwargs.find("movementThreat");
    if(it == kwargs.end())
      {
        return 0.0;
      }
    const auto *movementThreat = std::any_cast<const std::vector<double> *>(it->second);
    if(movementThreat == nullptr || movementThreat->empty())
      {
        return 0.0;
      }

    Combatant *combatant = _factory.getCombatant();
    const int lastIdx = static_cast<int>(movementThreat->size()) - 1;
    const int movement = combatant->getMovement();
    const int speed = combatant->getSpeed();

    if(_abilityType == AbilityType::AGGRESSIVE)
      {
        // Aggressive movement (Barbarian): we always want it when the destination can't be reached with the
        // current movement budget (the "move towards an enemy" intent is assumed), otherwise it is discouraged.
        return lastIdx > movement ? 1.0 : -1.0;
      }

    // Defensive dash: compare the threat we'd be exposed to stopping at our normal movement limit against the
    // threat at movement + speed (one extra Dash of distance) further along the same path. Only credit the
    // reduction (max(0, ...)); we don't want Dash to weigh in when it would merely carry us deeper into danger.
    const double baseline = -1.0 * (*movementThreat)[std::min(movement, lastIdx)];
    const double modified = -1.0 * (*movementThreat)[std::min(movement + speed, lastIdx)];
    return std::max(0.0, baseline - modified);
  }
}
